### PR TITLE
🐛 Adjust structuredmerge patch helper options to set correct allow list for Cluster objects to prevent co-ownership

### LINKED ISSUE #6736: Topology controller should not take ownership of cluster.spec.topology
**What steps did you take and what happened:**

Checked output of `pull-cluster-api-e2e-full-main`:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/directory/pull-cluster-api-e2e-full-main/1540324488158121984

Especially the cluster object and the resulted managed fields, the interesting part:
```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"cluster.x-k8s.io/v1beta1","kind":"Cluster","metadata":{"annotations":{},"labels":{"cni":"k8s-upgrade-and-conformance-l9e4bu-crs-0"},"name":"k8s-upgrade-and-conformance-l9e4bu","namespace":"k8s-upgrade-and-conformance-9sta2j"},"spec":{"clusterNetwork":{"pods":{"cidrBlocks":["192.168.0.0/16"]},"serviceDomain":"cluster.local","services":{"cidrBlocks":["10.128.0.0/12"]}},"topology":{"class":"quick-start","controlPlane":{"metadata":{},"replicas":3},"variables":[{"name":"etcdImageTag","value":""},{"name":"coreDNSImageTag","value":""}],"version":"v1.23.6","workers":{"machineDeployments":[{"class":"default-worker","failureDomain":"fd4","name":"md-0","replicas":1}]}}}}
  creationTimestamp: "2022-06-24T13:52:55Z"
  finalizers:
  - cluster.cluster.x-k8s.io
  generation: 4
  labels:
    cluster.x-k8s.io/cluster-name: k8s-upgrade-and-conformance-l9e4bu
    cni: k8s-upgrade-and-conformance-l9e4bu-crs-0
    topology.cluster.x-k8s.io/owned: ""
  managedFields:
  - apiVersion: cluster.x-k8s.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          f:cluster.x-k8s.io/cluster-name: {}
          f:cni: {}
          f:topology.cluster.x-k8s.io/owned: {}
      f:spec:
        f:clusterNetwork:
          f:pods:
            f:cidrBlocks: {}
          f:serviceDomain: {}
          f:services:
            f:cidrBlocks: {}
        f:controlPlaneRef:
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
          f:namespace: {}
        f:infrastructureRef:
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
          f:namespace: {}
        f:topology: // THE INTERESTING PART
          f:class: {}
          f:controlPlane:
            f:metadata: {}
            f:replicas: {}
          f:variables: {}
          f:version: {}
          f:workers:
            f:machineDeployments: {}
    manager: capi-topology
    operation: Apply
    time: "2022-06-24T14:12:06Z"
  - apiVersion: cluster.x-k8s.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          .: {}
          f:cni: {}
      f:spec:
        .: {}
        f:clusterNetwork:
          .: {}
          f:pods:
            .: {}
            f:cidrBlocks: {}
          f:serviceDomain: {}
          f:services:
            .: {}
            f:cidrBlocks: {}
        f:topology:
          .: {}
          f:class: {}
          f:controlPlane:
            .: {}
            f:metadata: {}
            f:replicas: {}
          f:workers:
            .: {}
            f:machineDeployments: {}
    manager: kubectl-client-side-apply
    operation: Update
    time: "2022-06-24T13:52:55Z"
```

[[0]](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6708/pull-cluster-api-e2e-full-main/1540324488158121984/artifacts/clusters/bootstrap/resources/k8s-upgrade-and-conformance-9sta2j/Cluster/k8s-upgrade-and-conformance-l9e4bu.yaml)

The interesting part here is the co-ownership of the path `spec.topology` and fields bellow.
These are fields which are normally read as input for computing the desired state and are normally never written by the topology controller.

The same may apply to the following paths but need to be verified:
* `.spec.paused`
* `spec.clusterNetwork`

The issue is: if the topology controller reconciles and at the same time another controller or user changes one of the co-owned fields, the topology controller may overwrite the changed data.

E.g.:
* topology controller starts reconciling cluster object, using the old values (`Cluster.spec.topology.workers.machineDeployments[0].replicas == 3`)
* user edits cluster and sets `Cluster.spec.topology.workers.machineDeployments[0].replicas = 100` and patches it to the api server
* topology controller finishes reconciliation which include spec changes, by that resets the cluster object to `Cluster.spec.topology.workers.machineDeployments[0].replicas == 3`

**What did you expect to happen:**

Topology controller does not overwrite cluster fields which it is not supposed to write to.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

1. Possible solution: Note: does not work out of the box, would require bigger refactoring in `patchEngine.Apply`
  * adjust computeCluster in topology controllers `desired_state.go` : 
As alternative `computeCluster` at https://github.com/kubernetes-sigs/cluster-api/blob/5c44a14256af7ce7cd6448b0987d83a87fb84d87/internal/controllers/topology/cluster/desired_state.go#L406

    to only contain set the relevant paths supposed to get set via topology controller:
    * `metadata.labels`
    * `cluster.Spec.InfrastructureRef`
    * `cluster.Spec.ControlPlaneRef`
    
    as well ass other required metadata
    * `metadata.name`
    * `metadata.namespace`

2. Possible solution:
  This could be adjusted by adding the unrelated fields to the ignorePaths at:
  * https://github.com/kubernetes-sigs/cluster-api/blob/5c44a14256af7ce7cd6448b0987d83a87fb84d87/internal/controllers/topology/cluster/reconcile_state.go#L442

  3. Possible solutions: other variants

xref: issue discovered while testing dry-run server side apply at #6710 but should potentially already affect main.

**Environment:**

- Cluster-api version:
- minikube/kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Prior to this fix, the topology controller added all spec fields of the incoming Cluster object to the intent to patch.
This lead to the capi-topology controller / manager taking ownership of fields it should not have an intention to set / should never set because they are changed by e.g. the user or git ops mechanisms, etc.

This resulted in the topology controller taking co-ownership of the fields (due to server side apply) and results in the following race condition:

* Topology Controller starts reconciling a Cluster
* Gets the current Cluster object (from its cache)
* Meanwhile an external change happens, e.g. changing the replicas of KCP from 3 to 5
* Topology Controller finishes reconciliation and patches the object, which also patches the replicas of KCP back from 5 to 3

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6736

### CODE REVIEW

**Comment by chrischdi** [comment]
Does not work out

/close

**Comment by sbueringer** [comment]
Sounds like a really serious bug

**Comment by chrischdi** [comment]
@sbueringer : this looks great 👍 

My favourite would be this approach instead of #6739

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 1]
```diff
@@ -440,7 +440,9 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, s *scope.Scope) error
 
 	// Check differences between current and desired state, and eventually patch the current object.
 	patchHelper, err := r.patchHelperFactory(s.Current.Cluster, s.Desired.Cluster, structuredmerge.IgnorePaths{
-		{"spec", "controlPlaneEndpoint"}, // this is a well known field that is managed by the Cluster controller, topology should not express opinions on it.
+		// This is a well known field that is managed by the Cluster controller, topology should not express opinions on it.
+		// It needs to be added here because it is not a pointer at the struct and would otherwise get defaulted by the topology controller.
+		{"spec", "controlPlaneEndpoint"},
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state.go:445]: This means we still have to add ignorePaths for all the non-pointer fields which have a similar issue?

I also prefer this variant, but that we have to add spec.controlPlaneEndpoint makes me wonder if we should use allowPaths additionally to make it future proof

Comment (sbueringer) [reply internal/controllers/topology/cluster/reconcile_state.go:445]: I just think this might end badly if we add new fields to the Cluster and don't remember this issue, which is why I would prefer having allowPaths here.

We could keep it simple and just export allowPaths and use them instead of hard-coded ones if the option is set.

Comment (sbueringer) [reply internal/controllers/topology/cluster/reconcile_state.go:445]: > At the end this should result in failing tests.

This assumes that we have e2e test coverage for this field

Comment (sbueringer) [reply internal/controllers/topology/cluster/reconcile_state.go:445]: Missed that we have care about some labels as well :joy:

File: internal/controllers/topology/cluster/desired_state.go [hunk 2]
```diff
@@ -404,21 +404,25 @@ func (r *Reconciler) computeControlPlaneVersion(ctx context.Context, s *scope.Sc
...
 // we are concerned only about the latest group of fields.
 func computeCluster(_ context.Context, s *scope.Scope, infrastructureCluster, controlPlane *unstructured.Unstructured) *clusterv1.Cluster {
-	cluster := s.Current.Cluster.DeepCopy()
-
-	// Enforce the topology labels.
-	// NOTE: The cluster label is added at creation time so this object could be read by the ClusterTopology
-	// controller immediately after creation, even before other controllers are going to add the label (if missing).
-	if cluster.Labels == nil {
-		cluster.Labels = map[string]string{}
+	cluster := &clusterv1.Cluster{
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/desired_state.go:407]: Computing a trimmed down desired state won't alleviate us from the need to get allowedPaths in order, due to json serialization weirdness, so I suggest moving this change to a separated PR/out of the critical path.
Let's keep this PR a small as possible

File: internal/controllers/topology/cluster/reconcile_state.go (hunk 1 above)
Comment (chrischdi) [reply internal/controllers/topology/cluster/reconcile_state.go]: I adjusted the helper.

I added an additional option which allows pass a more specific list instead of `{"spec"}` to handle the cluster use case.

File: internal/controllers/topology/cluster/desired_state.go (hunk 2 above)
Comment (chrischdi) [reply internal/controllers/topology/cluster/desired_state.go]: I reverted this changes on the PR 👍

**Comment by chrischdi** [comment]
/retitle 🐛 Add SSA patch helper option to specify allow list for spec fields to scope down Cluster objects

File: internal/controllers/topology/cluster/structuredmerge/options.go [hunk 3]
```diff
@@ -20,6 +20,23 @@ import (
...
+		{"kind"},
+		{"metadata", "name"},
+		{"metadata", "namespace"},
+		// the topology controller controls/has an opinion for labels, annotation, ownerReferences and spec only.
+		{"metadata", "labels"},
+		{"metadata", "annotations"},
+		{"metadata", "ownerReferences"},
+	}
+	// allowedPathSpec is the default allowed path set for the spec.
+	allowedPathSpec = contract.Path{"spec"}
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/structuredmerge/options.go:37]:
```suggestion
	defaultAllowedPathSpec = contract.Path{"spec"}
```

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 4]
```diff
@@ -439,8 +439,9 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, s *scope.Scope) error
 	ctx, log := tlog.LoggerFrom(ctx).WithObject(s.Desired.Cluster).Into(ctx)
 
 	// Check differences between current and desired state, and eventually patch the current object.
-	patchHelper, err := r.patchHelperFactory(s.Current.Cluster, s.Desired.Cluster, structuredmerge.IgnorePaths{
-		{"spec", "controlPlaneEndpoint"}, // this is a well known field that is managed by the Cluster controller, topology should not express opinions on it.
+	patchHelper, err := r.patchHelperFactory(s.Current.Cluster, s.Desired.Cluster, structuredmerge.AllowedSpecPaths{
+		{"spec", "infrastructureRef"},
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/reconcile_state.go:443]: Let's comment why for the Cluster we override default spec allowed paths

File: internal/controllers/topology/cluster/structuredmerge/options.go [hunk 5]
```diff
@@ -60,3 +77,13 @@ type IgnorePaths []contract.Path
...
 }
+
+// AllowedSpecPaths instruct the Helper to ignore everything in spec except given paths when computing a patch.
+// NOTE: ignorePaths are used to filter out fields nested inside allowedPaths, e.g.
+// spec.ControlPlaneEndpoint.
+type AllowedSpecPaths []contract.Path
+
+// ApplyToHelper applies this configuration to the given helper options.
+func (a AllowedSpecPaths) ApplyToHelper(opts *HelperOptions) {
+	opts.allowedPaths = append(allowedPathsMetadata, a...)
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/structuredmerge/options.go:88]: adding `allowedPathsMetadata` when processing `AllowedSpecPaths` is kind of confusing. At least we should document it better, but this is kind of hard to follow with the defaulting implemented in serversidepathhelper.go

File: internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go [hunk 6]
```diff
@@ -44,17 +44,8 @@ type serverSidePatchHelper struct {
...
-		{"kind"},
-		{"metadata", "name"},
-		{"metadata", "namespace"},
-		// the topology controller controls/has an opinion for labels, annotation, ownerReferences and spec only.
-		{"metadata", "labels"},
-		{"metadata", "annotations"},
-		{"metadata", "ownerReferences"},
-		{"spec"},
+	if len(helperOptions.allowedPaths) == 0 {
+		helperOptions.ApplyOptions([]HelperOption{AllowedSpecPaths{allowedPathSpec}})
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:48]: This is kind of hard to follow, given that it is meant to trigger also adding metadata allowed paths

File: internal/controllers/topology/cluster/structuredmerge/options.go [hunk 7]
```diff
@@ -18,6 +18,44 @@ package structuredmerge
...
+		{"metadata", "name"},
+		{"metadata", "namespace"},
+		// the topology controller controls/has an opinion for labels, annotation, ownerReferences and spec only.
+		{"metadata", "labels"},
+		{"metadata", "annotations"},
+		{"metadata", "ownerReferences"},
+		{"spec"},
+	}
+
+	// defaultAllowedPathsCluster are the allowed specific for Clusters.
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/structuredmerge/options.go:41]:
```suggestion
	// allowedPathsCluster are the allowed paths specific for Clusters.
```
I would drop the default as I would only consider the ones above as default now given that a consumer of the helper can't overwrite it anymore

**Comment by chrischdi** [comment]
/retitle 🐛 Adjust structuredmerge patch helper options to set correct allow list for Cluster objects to prevent co-ownership

/test all

**Review by sbueringer — COMMENTED** [review]
one nit

+/- I assume tested that it works :)

**Comment by sbueringer** [comment]
I resolved the comments above which don't apply anymore to the current implementation
